kong/runloop/plugin_servers/init.lua

```diff
@@ -0,0 +1,430 @@
+    ["kong.client.authenticate"] = "plugin.StepCredential",
+    ["kong.node.get_memory_stats"] = "plugin.StepMemoryStats",
+    ["kong.router.get_route"] = "plugin.StepRoute",
+    ["kong.router.get_service"] = "plugin.StepService",
+    ["kong.request.get_query"] = "plugin.StepMultiMap",
+    ["kong.request.get_headers"] = "plugin.StepMultiMap",
+    ["kong.response.get_headers"] = "plugin.StepMultiMap",
+    ["kong.service.response.get_headers"] = "plugin.StepMultiMap",
+  }
+
+  function get_step_method(step_in, pdk_res, pdk_err)
+    if not pdk_res and pdk_err then
+      return "plugin.StepError", pdk_err
+    end
+
+    return ((type(pdk_res) == "table" and pdk_res._method)
+      or by_pdk_method[step_in.Data.Method]
+      or "plugin.Step"), pdk_res
+  end
+end
+
+
+
+--[[
+
+--- Event loop -- instance reconnection
+
+--]]
+
+local function bridge_loop(instance_rpc, instance_id, phase)
+  if not instance_rpc then
+    kong.log.err("no instance_rpc: ", debug.traceback())
+  end
+  local step_in, err = instance_rpc:call("plugin.HandleEvent", {
+    InstanceId = instance_id,
+    EventName = phase,
+  })
+  if not step_in then
+    return step_in, err
+  end
+
+  local event_id = step_in.EventId
+
+  while true do
+    if step_in.Data == "ret" then
+      break
+    end
+
+    local pdk_res, pdk_err = call_pdk_method(
+      step_in.Data.Method,
+      step_in.Data.Args)
+
+    local step_method, step_res = get_step_method(step_in, pdk_res, pdk_err)
+
+    step_in, err = instance_rpc:call(step_method, {
+      EventId = event_id,
+      Data = step_res,
+    })
+    if not step_in then
+      return step_in, err
+    end
+  end
+end
+
+
+local function handle_event(instance_rpc, plugin_name, conf, phase)
+  local instance_id = get_instance_id(plugin_name, conf)
+  local _, err = bridge_loop(instance_rpc, instance_id, phase)
+
+  if err then
+    kong.log.err(err)
+
+    if string.match(err, "No plugin instance") then
+      reset_instance(plugin_name, conf)
+      return handle_event(instance_rpc, plugin_name, conf, phase)
+    end
+  end
+end
+
+
+--- Phase closures
+local function build_phases(plugin)
+  if not plugin then
+    return
+  end
+
+  local server_rpc = get_server_rpc(plugin.server_def)
+
+  for _, phase in ipairs(plugin.phases) do
+    if phase == "log" then
+      plugin[phase] = function(self, conf)
+        local saved = {
+          serialize_data = kong.log.serialize(),
+          ngx_ctx = ngx.ctx,
+          ctx_shared = kong.ctx.shared,
+        }
+
+        ngx_timer_at(0, function()
+          local co = coroutine.running()
+          save_for_later[co] = saved
+
+          handle_event(server_rpc, self.name, conf, phase)
+
+          save_for_later[co] = nil
+        end)
+      end
+
+    else
+      plugin[phase] = function(self, conf)
+        handle_event(server_rpc, self.name, conf, phase)
+      end
+    end
+  end
+
+  return plugin
+end
+
+
+
+--- module table
+local external_plugins = {}
```

Comment:
Given that the filename is `plugin_servers`, perhaps this should be called `plugin_servers` as well?